airbyte-integrations/connectors/source-google-adwords-singer/source_google_adwords_singer/source.py

```diff
@@ -23,38 +23,136 @@
 """
 
 import json
-import os
+import sys
+from typing import Dict, List
 
-from airbyte_protocol import AirbyteCatalog, AirbyteConnectionStatus, Status
-from base_python import AirbyteLogger, CatalogHelper
-from base_singer import SingerHelper, SingerSource
+from airbyte_protocol import AirbyteConnectionStatus, Status
+from base_python import AirbyteLogger
+from base_singer import SingerSource, SyncMode, SyncModeInfo
+from googleads import adwords, oauth2
+from tap_adwords import VERSION
 
 
 class SourceGoogleAdwordsSinger(SingerSource):
+    @staticmethod
+    def _get_accounts(logger: AirbyteLogger, sdk_client: adwords.AdWordsClient, selector: Dict):
+        # obtaining accounts for customer_id
+        managed_customer_page = sdk_client.GetService(service_name="ManagedCustomerService", version=VERSION).get(selector)
+        accounts = managed_customer_page.entries
+        return accounts
+
+    def _check_internal(self, logger: AirbyteLogger, streams: List, config: json):
+        # checking if REPORT syncing will be called for manager account
+        # https://developers.google.com/adwords/api/docs/common-errors#ReportDefinitionError.CUSTOMER_SERVING_TYPE_REPORT_MISMATCH
+        try:
+            customer_ids = config["customer_ids"].split(",")
+            oauth2_client = oauth2.GoogleRefreshTokenClient(
+                config["oauth_client_id"], config["oauth_client_secret"], config["refresh_token"]
+            )
+            for customer_id in customer_ids:
+                sdk_client = adwords.AdWordsClient(
+                    config["developer_token"], oauth2_client, user_agent=config["user_agent"], client_customer_id=customer_id
+                )
+                selector = {
+                    "fields": ["Name", "CanManageClients", "CustomerId", "TestAccount", "DateTimeZone", "CurrencyCode"],
+                    "predicates": [
+                        {
+                            "field": "CustomerId",
+                            "operator": "IN",
+                            "values": [
+                                customer_id,
+                            ],
+                        }
+                    ],
+                }
+                accounts = self._get_accounts(logger, sdk_client, selector)
+                if accounts:
+                    account = accounts[0]
+                    is_manager = account.canManageClients
+                    for stream in streams:
+                        if stream.endswith("REPORT") and is_manager:
+                            logger.log_by_prefix(f"Unable to sync {stream} with the manager account {customer_id}", "ERROR")
+                            sys.exit(1)
+                else:
+                    logger.log_by_prefix(f"Unable to sync with the provided customer id {customer_id}", "ERROR")
```

Comment:
Can this log also specify a reason why it's failing? What should the user who sees this message do?